pkg/util/validation/limits.go

```diff
@@ -265,8 +269,108 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 }
 
+type extension struct {
+	name string
+	typ  reflect.Type
+}
+
+// registeredExtensions is the list of registered extensions
+var registeredExtensions []extension
+
+// newExtensionsOnlyLimitsConfig returns an interface{} value of a pointer to a struct of type:
+//
+//	struct {
+//	    ExtName1 T1 `yaml:"extname1"`
+//	    ...
+//	    ExtN TN `yaml:"extN"`
+//	    Throwaway map[string]interface{} `yaml:",inline"`
+//	}
+//
+// Where TN is the type of the registered extension, and extnameN is the name of it.
+func newExtensionsOnlyLimitsConfig() (any interface{}, getter func() map[string]interface{}) {
+	fields := make([]reflect.StructField, 0, len(registeredExtensions)+1)
+	for _, e := range registeredExtensions {
+		fields = append(fields, reflect.StructField{
+			Name: strings.ToTitle(e.name),
+			Type: e.typ,
+			Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s"`, e.name)),
+		})
+	}
+	fields = append(fields, reflect.StructField{
+		Name: "Throwaway",
+		Type: reflect.TypeOf(map[string]interface{}{}),
+		Tag:  `yaml:",inline"`,
+	})
+	typ := reflect.StructOf(fields)
+	newExtensions := reflect.New(typ).Interface()
+	return newExtensions, func() map[string]interface{} {
+		ext := map[string]interface{}{}
+		for i, e := range registeredExtensions {
+			ext[e.name] = reflect.ValueOf(newExtensions).Elem().Field(i).Interface()
+		}
+		return ext
+	}
+}
+
+// newLimitsConfigThrowingAwayExtensions returns an instance of a struct with following structure:
+//
+//		type limitsConfigThrowingAwayExtensions struct {
+//			IgnoredExtensionExt1 interface{} `yaml:"ext1"`
+//	     ...
+//			IgnoredExtensionExtN interface{} `yaml:"ext1"`
+//			*plain               `json:",inline"`
+//		}
+//
+// This makes the YAML parser ignore the extension fields, and still validate our plain config.
+func newLimitsConfigThrowingAwayExtensions(plainLimits interface{}) interface{} {
+	fields := make([]reflect.StructField, 0, len(registeredExtensions)+1)
+	for _, e := range registeredExtensions {
+		fields = append(fields, reflect.StructField{
+			Name: "IgnoredExtension" + strings.ToTitle(e.name),
+			Type: e.typ,
+			Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s"`, e.name)),
+		})
+	}
+	fields = append(fields, reflect.StructField{
+		Name:      "PlainLimits",
+		Anonymous: true,
+		Type:      reflect.TypeOf(plainLimits),
+		Tag:       `yaml:",inline"`,
+	})
+	typ := reflect.StructOf(fields)
+	cfg := reflect.New(typ).Interface()
+	reflect.ValueOf(cfg).Elem().Field(len(registeredExtensions)).Set(reflect.ValueOf(plainLimits))
+	return cfg
+}
+
+// RegisterExtensions registers the extensions type with given name
+// and returns a function to get the extensions from a *Limits instance.
+// The name will be used as yaml key to decode the extensions.
+func RegisterExtensions[E any](name string) func(*Limits) *E {
```

Comment:
The idea behind this is that a package that defines a `NewFeatureLimits` would do:

```

type NewFeatureLimits struct {
	Enabled bool `yaml"enabled"`
}

var GetFromLimits = RegisterExtensions[NewFeatureLimits]("new_feature")
```

and then just use their package-level functions to retrieve their limits from the `*Limits`.